lib/handler/throttle_resp.c

```diff
@@ -0,0 +1,139 @@
+#include <stdlib.h>
+#include "h2o.h"
+
+#ifndef HUNDRED_MS
+#define HUNDRED_MS 100
+#endif
+
+#ifndef ONE_SECOND
+#define ONE_SECOND 1000
+#endif
+
+typedef H2O_VECTOR(h2o_iovec_t) iovec_vector_t;
+
+typedef struct st_throttle_resp_t {
+    h2o_ostream_t super;
+    h2o_timeout_entry_t timeout_entry;
+    int64_t tokens;
+    size_t token_inc;
+    h2o_context_t *ctx;
+    h2o_req_t *req;
+    struct {
+        iovec_vector_t bufs;
+        int is_final;
+    } state;
+} throttle_resp_t;
+
+static void real_send(throttle_resp_t *self) {
+    /* a really simple token bucket implementation */
+    assert(self->tokens > 0);
+    size_t i, token_consume;
+
+    token_consume = 0;
+
+    for (i = 0; i < self->state.bufs.size; i++) {
+        token_consume += self->state.bufs.entries[i].len;
+    }
+
+    self->tokens -= token_consume;
+
+    h2o_ostream_send_next(&self->super, self->req, self->state.bufs.entries,
+                          self->state.bufs.size, self->state.is_final);
+    if (self->state.is_final)
+        h2o_timeout_unlink(&self->timeout_entry);
+}
+
+static inline void expand_buf(h2o_mem_pool_t *pool, iovec_vector_t *bufs, size_t count) {
+    assert(bufs->capacity < count);
+    h2o_vector_reserve(pool, bufs, count);
+    bufs->capacity = count;
+}
+
+static void add_token(h2o_timeout_entry_t *entry) {
+    throttle_resp_t *self = H2O_STRUCT_FROM_MEMBER(throttle_resp_t, timeout_entry, entry);
+
+    h2o_timeout_link(self->ctx->loop, &self->ctx->hundred_ms_timeout, &self->timeout_entry);
+    self->tokens += self->token_inc;
+
+    if (self->tokens > 0)
+        real_send(self);
+}
+
+static void on_send(h2o_ostream_t *_self, h2o_req_t *req, h2o_iovec_t *inbufs, size_t inbufcnt, int is_final) {
+    throttle_resp_t *self = (void *)_self;
+    size_t i;
+
+    /* I don't know if this is a proper way. */
+    if (self->state.bufs.capacity < inbufcnt) {
+        expand_buf(&req->pool, &self->state.bufs, inbufcnt);
+    }
```

Comment:
I think you can replace this if block with a single call to `h2o_vector_reserve(&req->pool, &self->state.bufs, inbufcnt)` and also remove the definition of `expand_buf`.

`h2o_vector_reserve` expands the number of slots in `entries`, and sets the number of slots available to `capacity`.